File: lab6/lab6_2.kt
abstract class Book(
  private var _title: String,
  private var _pages: Int
) {
  abstract fun displayInfo()

  fun getTitle(): String { return _title }
  fun getPages(): Int { return _pages }
}

class Novel(
  t: String,
  pag: Int,
  private var genre: String
) : Book(t, pag) {

  override fun displayInfo() {
    println("Roman: ${getTitle()}")
    println("Gen: $genre")
    println("Pagini: ${getPages()}\n")
  }
}

class Reader {
  fun readBook(book: Book) {
    println("Cititorul citeste: ${book.getTitle()}")
    println("Numarul de pagini citite: ${book.getPages()}")
  }
}

fun comparePages(b1: Book, b2: Book) {
  if(b1.getPages() > b2.getPages())
    println("'${b1.getTitle()}' are mai multe pagini decat '${b2.getTitle()}'.")
  else if(b1.getPages() < b2.getPages())
    println("'${b1.getTitle()}' are mai putine pagini decat '${b2.getTitle()}'.")
  else
    println("'${b1.getTitle()}' si '${b2.getTitle()}' au acelasi numar de pagini.")
}

fun main() {
  val dune = Novel("Dune", 412, "Science Fiction")
  val stapanulInelelor = Novel("Stapanul Inelelor", 1178, "Fantasy")

  val ptrBook1: Book = dune
  val ptrBook2: Book = stapanulInelelor

  ptrBook1.displayInfo()
  ptrBook2.displayInfo()

  comparePages(dune, stapanulInelelor)
  println()

  val reader = Reader()
  reader.readBook(dune)
}

File: lab6/lab6_2.cpp
#include <iostream>
#include <string>

using namespace std;

class Reader;

class Book {
public:
  Book(const string& t, int pag) : title(t), pages(pag) {}

  virtual void displayInfo() const = 0;

protected:
  string title;
  int pages;

  friend class Reader;
  friend void comparePages(const Book& b1, const Book& b2);
};

class Novel : public Book {
public:
  Novel(const string& t, int pag, const string& g)
    : Book(t, pag), genre(g) {}

  void displayInfo() const override {
    cout << "Roman: " << title << "\n";
    cout << "Gen: " << genre << "\n";
    cout << "Pagini: " << pages << "\n\n";
  }

private:
  string genre;
};

class Reader {
public:
  void readBook(const Book& book) {
    cout << "Cititorul citeste: " << book.title << "\n";
    cout << "Numarul de pagini citite: " << book.pages << "\n";
  }
};

void comparePages(const Book& b1, const Book& b2) {
  if(b1.pages > b2.pages)
    cout << "'" << b1.title << "' are mai multe pagini decat '" << b2.title << "'.\n";
  else if(b1.pages < b2.pages)
    cout << "'" << b1.title << "' are mai putine pagini decat '" << b2.title << "'.\n";
  else
    cout << "'" << b1.title << "' si '" << b2.title << "' au acelasi numar de pagini.\n";
}

int main() {
  Novel dune("Dune", 412, "Science Fiction");
  Novel stapanulInelelor("Stapanul Inelelor", 1178, "Fantasy");

  Book* ptrBook1 = &dune;
  Book* ptrBook2 = &stapanulInelelor;

  ptrBook1->displayInfo();
  ptrBook2->displayInfo();

  comparePages(dune, stapanulInelelor);
  cout << "\n";

  Reader reader;
  reader.readBook(dune);

  return 0;
}